### PR TITLE
ApiHub to support poison queues 

### PR DESCRIPTION
**Pull request by safihamid**
ApiHub to support poison queues when there are multiple unsuccessful function execution attempts.

### CODE REVIEW

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 1]
```diff
@@ -20,13 +21,16 @@ namespace Microsoft.Azure.WebJobs.Extensions.ApiHub
     [Singleton(Mode = SingletonMode.Listener)]
     internal class ApiHubListener : IListener
     {
+        private const int MaxFunctionExecutionRetryCount = 5;
         private const string HostContainerName = "azure-webjobs-hosts";
+        private const string PoisonQueueName = "azure-webjobs-apihub-poison";
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: Thinking you should follow the naming scheme for blob trigger https://azure.microsoft.com/en-us/documentation/articles/websites-dotnet-webjobs-sdk-storage-blobs-how-to/

E.g. **webjobs-apihubtrigger-poison**

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 2]
```diff
@@ -20,13 +21,16 @@ namespace Microsoft.Azure.WebJobs.Extensions.ApiHub
     [Singleton(Mode = SingletonMode.Listener)]
     internal class ApiHubListener : IListener
     {
+        private const int MaxFunctionExecutionRetryCount = 5;
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: I think you should make this a configuration knob on your api hub config object. For queues/blobs this is configurable at the host level

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 3]
```diff
@@ -105,24 +114,56 @@ public void Dispose()
...
-                    pollStatus = uri.AbsoluteUri;
+                    status = new ApiHubStatus();
                 }
-                else
+
+                if (status.RetryCount < MaxFunctionExecutionRetryCount)
                 {
-                    pollStatus = obj.ToString();
+                    status.RetryCount++;
+                    _trace.Error(string.Format("Function {0} failed to successfully proces file {1}. Number of retries: {2}. ", _functionName, apiHubFile.Path, status.RetryCount));
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: spelling errors

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 4]
```diff
@@ -105,24 +114,56 @@ public void Dispose()
...
+                {
+                    // The maximum retries for the function execution has reached. The file info will be added to the poison queue and the next poll status will be updated to skip the file.
+                    status.PollUrl = pollStatus;
+                    status.RetryCount = 0;
+                    _trace.Error(string.Format("Function {0} failed to successfully proces file {1} after max allowed retries of {2}. The file info will be moved to queue {3}.", _functionName, apiHubFile.Path, MaxFunctionExecutionRetryCount, PoisonQueueName));
 
-                // If function successfully completes then the next poll Uri will be logged in an Azure Blob.
-                await SetNextPollStatusAsync(pollStatus);
+                    await MoveToPoisonQueueAsync(apiHubFile);
+                    await SetNextPollStatusAsync(status);
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: If the status update fails after you've moved to the poison queue, you could get duplicates in the queue right? I realize these two operations can't be done as an atomic transaction.

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 5]
```diff
@@ -134,7 +175,34 @@ public void Dispose()
             }
         }
 
-        private async Task<string> GetLastPollStatusAsync()
+        private async Task MoveToPoisonQueueAsync(ApiHubFile apiHubFile)
+        {
+            CloudQueueClient queueClient = CloudStorageAccount.Parse(this._config.StorageConnectionString).CreateCloudQueueClient();
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: You should do this once on construction, or otherwise find a way to cache the queue client

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 6]
```diff
@@ -134,7 +175,34 @@ public void Dispose()
...
+
+            var queue = queueClient.GetQueueReference(PoisonQueueName);
+
+            await queue.CreateIfNotExistsAsync();
+
+            var fileStatus = new ApiHubFileInfo
+            {
+                FilePath = apiHubFile.Path,
+                FunctionName = this._functionName,
+                Connection = this._connectionStringSetting
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: I think the connection string is resolved at this point? You want to avoid writing the actual connection string to the queue

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs [hunk 7]
```diff
@@ -134,7 +175,34 @@ public void Dispose()
...
-        private async Task<string> GetLastPollStatusAsync()
+        private async Task MoveToPoisonQueueAsync(ApiHubFile apiHubFile)
+        {
+            CloudQueueClient queueClient = CloudStorageAccount.Parse(this._config.StorageConnectionString).CreateCloudQueueClient();
+
+            var queue = queueClient.GetQueueReference(PoisonQueueName);
+
+            await queue.CreateIfNotExistsAsync();
+
+            var fileStatus = new ApiHubFileInfo
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: Has this stuff been released yet? This would be a breaking change if so.

**Comment by mathewc** [comment]
Need to add tests

File: src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs (hunk 6 above)
Comment (safihamid) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: It's not the actual connection string. It's the connectionStringSetting aka the key.

Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: Can you check its runtime value? Most of the binding attributes now use the new [AutoResolve] attribute which causes these values to be resolved automatically. Perhaps you're not using that and resolving manually. @brettsam might have comments as well.

Comment (safihamid) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: Not sure, why is this needed? Am I missing something? This is just to let the user know which trigger failed. We are giving them the file path, function name and the connection string name which is something like DROPBOX_dropbox for them to know which connector was the file coming from which failed.

Comment (brettsam) [inline src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs]: It's only resolved if he's using [AutoResolve(AllowTokens=false)] on this property. I think ApiHub handles it differently and resolves these on their own.

---
In reply to: [70373036](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/118#discussion_r70373036) [](ancestors = 70373036)

File: src/WebJobs.Extensions.ApiHub/Config/ApiHubConfiguration.cs [hunk 8]
```diff
@@ -65,6 +68,25 @@ public TraceWriter Logger
             }
         }
 
+        /// <summary>
+        /// Gets or sets the number of times to try processing a file before moving it to the poison queue (where
+        /// possible).
+        /// </summary>
+        public int MaxFunctionExecutionRetryCount
```
Comment (mathewc) [inline src/WebJobs.Extensions.ApiHub/Config/ApiHubConfiguration.cs]: nit: recommend simplifying this to MaxRetryCount

**Comment by mathewc** [comment]
:ship:
